internal/stage_handshake.go

```diff
@@ -63,9 +60,29 @@ func testHandshake(stageHarness *test_case_harness.TestCaseHarness) error {
 		return err
 	}
 
+	expectedReservedBytes := []byte{0, 0, 0, 0, 0, 0, 0, 0}
```

Comment:
good question, yes, I was thinking people would have to build-in the capability to enable / disable extension support for older stages. would be a good thing to call out in the course description if we're going down this path

is there a better way? if we were to become more lenient for the handshake stage, say also accept extensions in handshake tests for the base challenge,  I think download piece and download file stages would also fail because the sequence of messages is different (you'd need to do the extension handshake once you signal support for extensions)

Reply:
Generally, we try to keep tests flexible enough to handle later implementations and don't burden the user with having to maintain backward compatibility.

I do see how this could interact with the download_piece/download_file though. Let me think about this a bit... will respond today

Reply:
@sarp a peer would only initiate the extension handshake **if** the base handshake received included the extensions bit, yes? In that case I think we're good - users won't have to disable/enable extension support based on the stage they're on (i.e. the command they're handling), they'd just need to do so based on what bits they receive from a peer in the initial handshake. 

We'd still want to be tolerant of the user sending their extension bit set in the initial handshake (for any stage), but we don't need to send it in our client's handshake for earlier stages - and thus won't have to expect the extension handshake to take place.

Reply:
> a peer would only initiate the extension handshake if the base handshake received included the extensions bit, yes? 

correct

> In that case I think we're good - users won't have to disable/enable extension support based on the stage they're on (i.e. the command they're handling), they'd just need to do so based on what bits they receive from a peer in the initial handshake.

problem is that the tracker you've setup on `bittorrent-test-tracker.codecrafters.io` always indicates during base handshake that it supports extensions, regardless of your client supporting it or not. and once both parties support it, extension handshake messages will be sent, also for base stages ...

I see two options:
1) one way could be to have two sets of trackers, one for base stages that is configured without extensions support and one tracker for magnet links and future challenges that need extensions
2) other way is that users build the capability to disable extensions for base stages and enable it for magnet links and future extensions

Reply:
What I'm trying to avoid here is anything "CodeCrafters-specific" - code a user would have to write only to pass CodeCrafters tests, but wouldn't have to write when building a real-world implementation. I think the flow described above doesn't count as CodeCrafters-specific - it's just something an implementation would have to handle anyway. 

(Maybe this is what you meant by "people would have to build-in the capability to enable / disable extension support for older stages", apologies if so)

Reply:
No, it seems reasonable to me to have the ability to enable / disable extension support in a BitTorrent client, wouldn't call it "CodeCrafters-specific"